valuetransfer/src/main/java/nl/tudelft/trustchain/valuetransfer/util/SecurityUtil.kt

```diff
@@ -0,0 +1,94 @@
+package nl.tudelft.trustchain.valuetransfer.util
+
+import android.os.Build
+import android.security.keystore.KeyGenParameterSpec
+import android.security.keystore.KeyProperties
+import androidx.annotation.RequiresApi
+import java.lang.RuntimeException
+import java.math.BigInteger
+import java.net.URLDecoder
+import java.net.URLEncoder
+import java.security.*
+import java.security.interfaces.RSAPublicKey
+import java.security.spec.RSAPublicKeySpec
+import javax.security.auth.x500.X500Principal
+
+object SecurityUtil {
+
+    private const val KEY_ALIAS = "nl.tudelft.trustchain.key"
+    private const val ANDROID_KEYSTORE = "AndroidKeyStore"
+
+    @RequiresApi(Build.VERSION_CODES.M)
+    fun generateKey(): KeyPair {
+        // We are creating a RSA key pair
+        val keyPairGenerator: KeyPairGenerator = KeyPairGenerator.getInstance(KeyProperties.KEY_ALGORITHM_RSA, ANDROID_KEYSTORE)
+        // We are creating the key pair with sign and verify purposes
+        val parameterSpec: KeyGenParameterSpec = KeyGenParameterSpec.Builder(
+            KEY_ALIAS,
+            KeyProperties.PURPOSE_SIGN or KeyProperties.PURPOSE_VERIFY
+        ).run {
+            setCertificateSerialNumber(BigInteger.valueOf(777)) // Serial number used for the self-signed certificate of the generated key pair, default is 1
+            setCertificateSubject(X500Principal("CN=$KEY_ALIAS")) // Subject used for the self-signed certificate of the generated key pair, default is CN=fake
+            setDigests(KeyProperties.DIGEST_SHA256) // Set of digests algorithms with which the key can be used
+            setSignaturePaddings(KeyProperties.SIGNATURE_PADDING_RSA_PKCS1) // Set of padding schemes with which the key can be used when signing/verifying
+            build()
+        }
+
+        // Initialization of key generator with the parameters we have specified above
+        keyPairGenerator.initialize(parameterSpec)
+
+        // Generates the key pair
+        return keyPairGenerator.genKeyPair()
+    }
+
+    fun serializePK(publicKey: RSAPublicKey): String {
+        return publicKey.modulus.toString() + "||" + publicKey.publicExponent.toString()
+    }
+
+    fun deserializePK(publicKey: String?): PublicKey? {
+        return try {
+            val split = publicKey.toString().split("||")
+            val spec = RSAPublicKeySpec(BigInteger(split[0]), BigInteger(split[1]))
+            KeyFactory.getInstance(KeyProperties.KEY_ALGORITHM_RSA).generatePublic(spec)
+        } catch (ex: Exception) {
+            null
+        }
+    }
+
+    fun sign(input: String, privateKey: PrivateKey): String {
```

Comment:
Would it be possible to use the underlying signature primitives provided by IPv8? If not, why not?

Reply:
Yes, added it to suggestions in the readme.